src/ray/object_manager/pull_manager.cc

```diff
@@ -0,0 +1,135 @@
+#include "ray/object_manager/pull_manager.h"
+
+namespace ray {
+
+PullManager::PullManager(
+    NodeID &self_node_id, const std::function<bool(const ObjectID &)> object_is_local,
+    const std::function<void(const ObjectID &, const NodeID &)> send_pull_request,
+    const std::function<int(int)> get_rand_int,
+    const RestoreSpilledObjectCallback restore_spilled_object)
+    : self_node_id_(self_node_id),
+      object_is_local_(object_is_local),
+      send_pull_request_(send_pull_request),
+      get_rand_int_(get_rand_int),
+      restore_spilled_object_(restore_spilled_object) {
+  object_is_local(ObjectID::FromRandom());
+}
+
+bool PullManager::Pull(const ObjectID &object_id, const rpc::Address &owner_address) {
+  RAY_LOG(DEBUG) << "Pull "
+                 << " of object " << object_id;
+  // Check if object is already local.
+  if (object_is_local_(object_id)) {
+    RAY_LOG(ERROR) << object_id << " attempted to pull an object that's already local.";
+    return false;
+  }
+  if (pull_requests_.find(object_id) != pull_requests_.end()) {
+    RAY_LOG(DEBUG) << object_id << " has inflight pull_requests, skipping.";
+    return false;
+  }
+
+  pull_requests_.emplace(object_id, PullRequest());
+  return true;
+}
+
+void PullManager::OnLocationChange(const ObjectID &object_id,
+                                   const std::unordered_set<NodeID> &client_ids,
+                                   const std::string &spilled_url) {
+  // Exit if the Pull request has already been fulfilled or canceled.
+  auto it = pull_requests_.find(object_id);
+  if (it == pull_requests_.end()) {
+    return;
+  }
+  // Reset the list of clients that are now expected to have the object.
+  // NOTE(swang): Since we are overwriting the previous list of clients,
+  // we may end up sending a duplicate request to the same client as
+  // before.
+  it->second.client_locations = std::vector<NodeID>(client_ids.begin(), client_ids.end());
+  if (!spilled_url.empty()) {
+    // Try to restore the spilled object.
+    restore_spilled_object_(object_id, spilled_url,
+                            [this, object_id](const ray::Status &status) {
+                              // Fall back to fetching from another object manager.
+                              if (!status.ok()) {
+                                TryPull(object_id);
+                              }
+                            });
+  } else if (it->second.client_locations.empty()) {
+    // The object locations are now empty, so we should wait for the next
+    // notification about a new object location.  Cancel the timer until
+    // the next Pull attempt since there are no more clients to try.
+    if (it->second.retry_timer != nullptr) {
+      it->second.retry_timer->cancel();
+      it->second.timer_set = false;
```

Comment:
It seems we never initialize the timer now, is this to be implemented?

Reply:
Hmm this timer can probably be cleaned up now that we will have the global timer